### ISSUE TITLE
can it use in ipc？

### ISSUE THREAD

**Issue by 0428402001** [cameron314/readerwriterqueue#126]
can it use in ipc？

**Comment by cameron314** [comment]
Not with shared memory, no.

**Comment by FloatingUpstream** [comment]
May I ask what the issue in the case of shared memory is?
My unerstanding: std::atomic is only going to work in a single process, just as _mm_mfence, etc.

How does Boost's spsc_queue solve this in combination with shared memory?
https://stackoverflow.com/questions/22207546/shared-memory-ipc-synchronization-lock-free/22209595#22209595

As far as I can tell, it will usually just default to std::atomic and memory_* orderings as well? Is the stackoverflow thread simply wrong?

Thanks for any clarifiation!

**Comment by cameron314** [comment]
Shared memory is not covered by the C++ standard (it's completely platform-dependent).

Setting that aside (even if the shared memory behaves "as expected" on your target platform and you're willing to accept the risk of undefined behaviour), the queue relies in part on thread-local addresses/IDs for implicit producers, which breaks when multiple processes are involved.

My queue also allocates memory dynamically, so you'd need to plug in a special `malloc`/`free` implementation via its traits to even attempt to use it between processes.

The blocking version uses various platform-specific semaphore primitives, some of which are not safe to place and manipulate in shared memory. On Linux, it appears that POSIX semaphores (used by the queue) are actually supported in shared memory between processes, but `1` needs to be passed to `sem_init` to enable this feature (currently `0` is passed since this presumably has a cost).

So in short, it *could* work if you try hard enough (and avoid implicit producers, blocking, and possibly other features), but there's no guarantee that it wouldn't break arbitrarily some day, and it's outside the scope of this project to support it.

**Comment by FloatingUpstream** [comment]
Many thanks for the explanation.
I guess I am out of luck in that case (especially targeting x64 and ARM). :-)

**Comment by cameron314** [comment]
Oops, mixed up which repo this is -- readerwriterqueue does not use thread-local addresses/IDs, but the rest still applies. (Although there are no traits to override `malloc`/`free`.)
